diff --git a/cli.c b/cli.c
index f8b45db..635639a 100644
--- a/cli.c
+++ b/cli.c
@@ -152,7 +152,7 @@ static void uci_usage(void)
 		"\treorder    <config>.<section>=<position>\n"
 		"\n"
 		"Options:\n"
-		"\t-c <path>  set the search path for config files (default: /etc/config)\n"
+		"\t-c <path>  set the search path for config files (default: /userdisk/miwifi/etc/config)\n"
 		"\t-d <str>   set the delimiter for list values in uci show\n"
 		"\t-f <file>  use <file> as input instead of stdin\n"
 		"\t-m         when importing, merge data into an existing package\n"
diff --git a/uci.h b/uci.h
index 43f2052..160d2a6 100644
--- a/uci.h
+++ b/uci.h
@@ -35,8 +35,8 @@ extern "C" {
 #include <stdio.h>
 #include <stdint.h>
 
-#define UCI_CONFDIR "/etc/config"
-#define UCI_SAVEDIR "/tmp/.uci"
+#define UCI_CONFDIR "/userdisk/miwifi/etc/config"
+#define UCI_SAVEDIR "/tmp/.uci_v3"
 #define UCI_DIRMODE 0700
 #define UCI_FILEMODE 0600
 
